src/activity_actor.cpp

```diff
@@ -387,11 +579,20 @@ void open_gate_activity_actor::finish( player_activity &act, Character & )
     act.set_to_null();
 }
 
+/** @pre @p other is an open_gate_activity_actor */
+bool open_gate_activity_actor::can_resume_with( const activity_actor &other,
+        const Character & ) const
+{
+    const open_gate_activity_actor &og_actor = dynamic_cast<const open_gate_activity_actor &>( other );
```

Comment:
I recall there is some sort of template trick to let overriden functions take the derived type instead of the base type as parameters, but I'm not clear how exactly that is done.

Reply:
I was hoping there was a way to do this and it would make the code a lot cleaner.

Reply:
I've been thinking more about it and possibly the alternative I mentioned in  "Describe alternatives you've considered" would be better.  I have edited the original post to reflect that.

Reply:
You *could* use [CRTP](https://en.wikipedia.org/wiki/Curiously_recurring_template_pattern) for this, but it would be complicated.  You'd have to introduce a new CRTP intermediate class to put the cast in, so the inheritance would be:
`open_gate_activity_actor` -> `activity_actor_crtp<open_gate_activity_actor>` -> `activity_actor`, and you could put that `dynamic_cast` in that intermediate class.  I'm not sure whether it's worth the extra complexity for just this; my instinct is probably not, unless you see some other `activity_actor` code that would also be simplified by this change.

Reply:
Thanks! I was not aware of that pattern--it seems pretty interesting.  I agree that it is probably overkill in this case, but I'd be happy to go with this approach if others think it would be a good way to do it.